libraries/CryptoInterface/src/CryptoInterface.h

```diff
@@ -0,0 +1,806 @@
+    constexpr uint8_t SHA1_NATURAL_LENGTH = 20;
+    constexpr uint8_t SHA224_NATURAL_LENGTH = 28;
+    constexpr uint8_t SHA256_NATURAL_LENGTH = 32;
+    constexpr uint8_t SHA384_NATURAL_LENGTH = 48;
+    constexpr uint8_t SHA512_NATURAL_LENGTH = 64;
+
+    /**
+     * MD5+SHA-1 is the concatenation of MD5 and SHA-1 computed over the same input; in the implementation, the internal data buffer is shared, 
+     * thus making it more memory-efficient than separate MD5 and SHA-1. It can be useful in implementing SSL 3.0, TLS 1.0 and TLS 1.1.
+     */
+    constexpr uint8_t MD5SHA1_NATURAL_LENGTH = 36;
+
+    constexpr uint8_t ENCRYPTION_KEY_LENGTH = 32;
+
+    constexpr uint32_t CT_MAX_DIFF = 1073741823; // 2^30 - 1
+
+    /**
+     * This function allows for fine-tuning of the specifications for the constant time calculations. 
+     * It should not be changed once a constant time function has been used at least once. 
+     * Otherwise the constant time will not be constant for the used functions.
+     * 
+     * The difference getCtMaxDataLength() - getCtMinDataLength() MUST be less than 2^30 (i.e. about one gigabyte).
+     */
+    void setCtMinDataLength(const size_t ctMinDataLength);
+    /**
+     * 0 by default.
+     */
+    size_t getCtMinDataLength();
+
+    /**
+     * This function allows for fine-tuning of the specifications for the constant time calculations. 
+     * It should not be changed once a constant time function has been used at least once. 
+     * Otherwise the constant time will not be constant for the used functions.
+     * 
+     * The difference getCtMaxDataLength() - getCtMinDataLength() MUST be less than 2^30 (i.e. about one gigabyte).
+     */
+    void setCtMaxDataLength(const size_t ctMaxDataLength);
+    /**
+     * 1024 by default.
+     */
+    size_t getCtMaxDataLength();
+
+    /**
+     * Set the nonce generator used by the CryptoInterface functions.
+     * 
+     * @param nonceGenerator The nonce generator to use.
+     */
+    void setNonceGenerator(nonceGeneratorType nonceGenerator);
+    nonceGeneratorType getNonceGenerator();
+    
+
+    // #################### MD5 ####################
+
+    /**
+     * WARNING! The MD5 hash is broken in terms of attacker resistance.
+     * Only use it in those cases where attacker resistance is not important. Prefer SHA-256 or higher otherwise.
+     * 
+     * Create a MD5 hash of the data. The result will be MD5_NATURAL_LENGTH bytes long and stored in resultArray.
+     * Uses the BearSSL cryptographic library.
+     * 
+     * @param data The data array from which to create the hash.
+     * @param dataLength The length of the data array in bytes.
+     * @param resultArray The array wherein to store the resulting hash. MUST be be able to contain MD5_NATURAL_LENGTH bytes or more.
+     * 
+     * @return A pointer to resultArray.
+     */
+    void *md5Hash(const void *data, const size_t dataLength, void *resultArray) __attribute__((deprecated));
```

Comment:
It seems to me that the string of hashes all have the same API (since they're just wrappers for BSSL functions and, well, hashes have a well defined interface).  Is this not a good spot for a *little* abstraction with a `HashClass` parent and a set of descendent objects (MD5, SHA1, etc.)?  You'd then call with `Crypto::MD5::hash(x....)` and get sizes for arrays with `uint8_t dest[MD5:hashSize];` and be able to pass them around abstractly (say, on user choice or handshake options) as `HashClass` or something?

Reply:
This is what I meant with my comment about something like struct md5 {} struct sha1 {} etc. It means restructuring.
A note here: if a parent class is to be implemented, I suggest doing that only as guidance for a common interface, and using non-virtualness in the parent. In other words, aim for templated usage and not polymorphism,

Reply:
There is some room for abstraction here, indeed. @devyte Do you have some existing code base as an example for the templated usage you would like? It would help me avoid misunderstandings.

Reply:
There's already a HashClass used in the BearSSLHelpers.h.  I know @devyte hates virtual methods, but IMO a nice virtual base and overridden methods is good enough and lets us have 1 copy of the code in flash that can handle any kind of hash (vs. templates where you regen a completely new class...which can't be used by others without also making them templated...it's a virus!)

Reply:
I was thinking just free standing structs, e. g. struct md5, struct sha1, struct sha256, etc, each with methods hash(), and whatever else is needed.
I somehow don't think they can all have the exact same interface, which would make polymorphism very messy, because the base class would need to have a superset of all cases. Let's try free standing objects first, and if needed we can implement a base class later and have all inherit from that.
About the BearSSL HashClass, it can be reworked as a thin polymorphic wrapper around these if needed.

Reply:
Code classified.